prefetch-npm-deps: add cacheVersion for packument support

Mic92 opened the pull request:
Add a cacheVersion parameter to fetchNpmDeps and npmDepsCacheVersion to buildNpmPackage. When set to 2, prefetch-npm-deps will also fetch and cache packuments (package metadata) in addition to tarballs.

npm can request packuments with two different Accept headers:
- corgiDoc: abbreviated metadata (default)
- fullDoc: full metadata (used for workspaces)

npm's cache policy requires headers to match, so we cache both versions.

This is opt-in via cacheVersion to avoid breaking existing hashes. Set npmDepsCacheVersion = 2 for projects using npm workspaces.

Also fix cacache index format to properly separate multiple entries with newlines, and update map_cache() to parse multi-line index files.


This change allows me to build qwen-code.

Fixes https://github.com/NixOS/nixpkgs/issues/474535

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Mic92 committed:
prefetch-npm-deps: add cacheVersion for packument support

Add a cacheVersion parameter to fetchNpmDeps and npmDepsCacheVersion to
buildNpmPackage. When set to 2, prefetch-npm-deps will also fetch and
cache packuments (package metadata) in addition to tarballs.

npm can request packuments with two different Accept headers:
- corgiDoc: abbreviated metadata (default)
- fullDoc: full metadata (used for workspaces)

npm's cache policy requires headers to match, so we cache both versions.

This is opt-in via cacheVersion to avoid breaking existing hashes.
Set npmDepsCacheVersion = 2 for projects using npm workspaces.

Also fix cacache index format to properly separate multiple entries
with newlines, and update map_cache() to parse multi-line index files.

Mic92 committed:
doc: add release note for buildNpmPackage packument support

Document the new npmDepsCacheVersion option that enables packument
caching for npm workspaces support.

winterqt commented:
Working through this now, there are some issues with the version of npm included in Node 24 I'm trying to work through myself before passing it off :P

winterqt committed:
buildNpmPackage: add diagnostic for cache version mismatch

winterqt committed:
npmConfigHook: always make cache writable when using cache v2

winterqt committed:
npmConfigHook: suggest cache v2 when `npm install` fails

winterqt reviewed (commented):
All in all, this looks good -- just a few comments/concerns. I pushed a few changes made when testing this to improve the UX, and ran rustfmt on your original commit (hence the force push). Thanks a ton!

Mic92 commented on pkgs/build-support/node/prefetch-npm-deps/src/main.rs:
```diff
@@ -22,6 +23,104 @@ fn cache_map_path() -> Option<PathBuf> {
+fn extract_package_name_from_url(url: &Url) -> Option<String> {
+    // Only handle npm registry URLs
+    let host = url.host_str()?;
+    if !host.contains("npmjs.org") && !host.contains("npm.") {
```
That was a clumsy attempt to also match private registries i.e.
`https://npm.pkg.github.com`, but we can also stick to npmjs.org for now.

winterqt replied:
Moot with https://github.com/NixOS/nixpkgs/pull/470517#discussion_r2654850776.

Mic92 replied:
Now that we are not using urls anymore, this is no longer needed at all.

Mic92 committed:
prefetch-npm-deps: extract package names from lockfile keys

The current approach parses tarball URLs to extract package names for
packument fetching. This is fragile as it only handles npmjs.org URLs
and requires special-casing other registries.

Use lockfile keys directly instead. The lockfile already contains the
canonical package names in the form "node_modules/@scope/name", so we
can simply strip the prefix rather than parsing URLs.

This handles all registries uniformly and eliminates the URL parsing
code along with its tests.

winterqt commented on pkgs/build-support/node/build-npm-package/hooks/npm-config-hook.sh:
```diff
@@ -78,7 +101,11 @@ npmConfigHook() {
 
-    if [ -z "${makeCacheWritable-}" ]; then
+    # When a given cache key has multiple entries (which is the case with
+    # cache version 2), npm always needs to write to the cache.
```
It is based on my reading of the npm logic, I just wanted to make sure. Thanks.

Mic92 commented on pkgs/build-support/node/prefetch-npm-deps/src/main.rs:
```diff
@@ -22,6 +23,104 @@ fn cache_map_path() -> Option<PathBuf> {
+/// Extract the package name from an npm registry tarball URL.
+/// e.g., "https://registry.npmjs.org/@types/react-dom/-/react-dom-19.1.6.tgz"
+///    -> Some("@types/react-dom")
+fn extract_package_name_from_url(url: &Url) -> Option<String> {
```
Done.

winterqt commented on pkgs/build-support/node/prefetch-npm-deps/src/cacache.rs:
```diff
@@ -115,13 +123,27 @@ impl Cache {
+            .open(&index_path)?;
+
+        // cacache format uses newline as entry separator (see cacache entry-index.js)
+        // Only add newline prefix if file already has content, to maintain backwards compatibility
```
```suggestion
        // Only add newline prefix if file already has content, to maintain backwards compatibility with previous versions of prefetch-npm-deps, which handled this case incorrectly.
```
for posterity of what we’re maintaining back compat with.

(I’m very willing to always write newline prefix for cache ver >= 2, btw, though you’d have to plumb the version through and idk if that’s worth it at this point.)

Mic92 commented on pkgs/build-support/node/prefetch-npm-deps/src/main.rs:
```diff
@@ -241,6 +330,14 @@ fn main() -> anyhow::Result<()> {
+    // Extract from lockfile keys like "node_modules/@scope/name" -> "@scope/name"
+    let package_names: HashSet<String> = packages
+        .iter()
+        .filter_map(|p| p.name.rsplit_once("node_modules/").map(|(_, n)| n.to_string()))
```
Look like this is not working for qwen-code yet. I get a cache mis-match:

```
$ cat /nix/store/rdyv89p30ksn5cickws6apqgq7i173h9-source/package-lock.json | jq -r '.packages | keys[]' | grep react-dom | head -5
  ⎿  node_modules/@types/react-dom
     node_modules/react-dom
     packages/vscode-ide-companion/node_modules/@types/react-dom
     packages/vscode-ide-companion/node_modules/react-dom
```

Mic92 replied:
Adding back the makeCacheWritable indeed fixed it.

Mic92 committed:
prefetch-npm-deps: clarify backwards compatibility comment

Mic92 commented:
Here is the diff for qwen-code:

```
  File counts:
  - Darwin: 4853 files
  - Linux: 4855 files (2 extra)

  Extra files on Linux:
  1. _cacache/content-v2/sha512/1a/b3/... - packument JSON for string-width-cjs
  2. _cacache/index-v5/9b/53 - cache index entry for string-width-cjs

```

Mic92 committed:
prefetch-npm-deps: use package name field for aliases

npm lockfiles can contain package aliases where the lockfile key differs
from the actual package name (e.g., "string-width-cjs" aliasing
"string-width"). Previously we always used the lockfile key, causing us
to fetch packuments for the wrong package.

Use the package's own "name" field when present, falling back to the
lockfile key. This ensures we fetch the correct packument for aliased
packages, fixing non-deterministic builds where the wrong packument
fetch could succeed or fail depending on network timing.

SuperSandro2000 commented on pkgs/build-support/node/build-npm-package/hooks/npm-config-hook.sh:
```diff
@@ -28,6 +28,29 @@ npmConfigHook() {
+
+    # Only run this in buildNpmPackage, this is just for a nicer error message; we trust that
+    # people using the setup hook directly also know how FODs work. ;)
+    if [[ -n ${NIX_NPM_CACHE_VERSION+x} ]] && [[ $NIX_NPM_CACHE_VERSION != $cacheVersion ]]; then
```
```suggestion
    if [[ -n ${NIX_NPM_CACHE_VERSION+x} && $NIX_NPM_CACHE_VERSION != $cacheVersion ]]; then
```

Mic92 committed:
fetchNpmDeps: rename cacheVersion to fetcherVersion

To align with pnpm tooling.

Mic92 committed:
prefetch-npm-deps: normalize packuments for deterministic hashes

Fixed-output derivations for npm dependencies fail with hash mismatches
when rebuilt on different days. The root cause is that npm registry
metadata (packuments) contain volatile fields that change whenever
upstream publishes a new version.

For example, the TypeScript packument changes daily due to nightly
releases, even though the lockfile pins a specific version. The cached
packument includes _rev, time, modified, and the full versions list -
all of which drift over time.

Strip these volatile fields and filter the versions map to only include
versions actually referenced in the lockfile.

Signed-off-by: Jörg Thalheim <joerg@thalheim.io>

Mic92 commented:
> Is it possible to exclude the FODs from your cache so we can make sure that the reproducibility is sane by having all of the users build it?

There are not included in the cache. The final package is though. Based on past experience there are plenty users that do not use the cache as for most packages the build time is not that high.

Mic92 committed:
prefetch-npm-deps: normalize packuments using whitelist approach

Switch from stripping known volatile fields to using an explicit whitelist
of allowed fields. This is more robust against upstream changes that add
new fields which could affect hash stability.

Top-level: only name and versions (dist-tags and time not needed for
lockfile installs where versions are already resolved)

Version-level: identity, all dependency types, dist, bin, platform
constraints (engines/os/cpu), scripts, and deprecated flag.

Based on analysis of pacote, npm-pick-manifest, npm-install-checks, and
arborist - only fields actually read during npm install are included.

niklaskorz commented:
Do you think this would be acceptable to backport to 25.11? We have several security updates in a package that do not build without the new fetcher version.
